### ISSUE TITLE
Ganache should not mine multiple blocks with the same timestamp.

### ISSUE THREAD

**Issue by MicahZoltu** [trufflesuite/ganache#1475]
Part of the Ethereum consensus rules are that each block must have a timestamp that is > the timestamp of its parent block.  Currently, ganache will sometimes mine a block with a timestamp that is == to its parent block, which violates consensus.

This consensus violation can result in tests against ganache failing when they wouldn't fail against a real network if the contract expects timestamps to be strictly increasing between blocks.

## Possible Solution
When ganache mines a block, the timestamp should always be > the parent block's timestamp, even if that puts the block in the future (actual timestamp doesn't matter, what matters is the strictly increasing aspect).

An exception could be made if the user explicitly sets the timestamp (via `evm_mine` or `evm_adjustTime` calls) since this is basically the user saying, "trust me, I know what I'm doing".

**Comment by davidmurdoch** [comment]
Keep in mind that Ganache itself does not _need_ consensus internally. Its main purpose to to make development easier.

So I don't think I agree here. Currently, Ethereum cannot mine two blocks within the same second; it's just not technically possible for the network to so. And Yellow Paper does forbid it. However, there may be a future where same-second block times _are_ possible, albeit this the prospect is slim to none.

Writing contract code that relies on sequential timestamps (timestamp that is > the timestamp of its parent block) is similar to writing contract code relies on opcode gas costs. These sorts of things could change in the future (though I do doubt it ever will, in this case).

Ganache being able to mine blocks within the same millisecond is possible. Insta-mining is a huge benefit to using Ganache. Users _do_ expect timestamps to be accurate. Currently, the plan is to keep timestamps as they are, but maybe we can add an option/flag (`strictConsensusRules`, maybe?) that enabled opting in to always increasing block timestamps (and any other rules Ganache may intentionally break for developmental reasons).

**Comment by MicahZoltu** [comment]
> Writing contract code that relies on sequential timestamps (timestamp that is > the timestamp of its parent block) is similar to writing contract code relies on opcode gas costs. These sorts of things could change in the future (though I do doubt it ever will, in this case).

I disagree with this pretty strongly, and this is coming from one of the foremost advocates against depending on gas costs in transactions.  Gas costs are, by design, meant to change over time.  The assertion that every block will have a timestamp that is strictly increasing is pretty core to how Ethereum (and proof of work chains in general) function.  Technically, _anything_ can change, including the fact that blocks have hashes (for example).  However, in order to do useful things we must be able to make some reasonable assertions about how the world will work in the future and I think that "time always goes up" is one such reasonable assertion.

I would support an optional flag for this, though I dislike bundling a flag with a bunch of other things since I may benefit from some aspects of violating consensus while I do not benefit from others.

> Users do expect timestamps to be accurate

Originally when I read this I was convinced that this was a strong argument for making this an option rather than default/enforced, but the more I think about it the more I disagree with it.  Timestamps are well known to _not_ be accurate on a fine grained level, they are only accurate over sufficiently long time spans.  Contracts **MUST** not have a hard dependency on timestamps being within seconds of accurate, and even minutes is incredibly dubious (I have seen miners lie about timestamps by quite a bit in the real world).

I think Ganache should, by default, publish blocks with a timestamp of `Math.max(now, parentBlock.timestamp + 1)`.  Such a strategy would result in timestamps usually being pretty close (unless you are generating over than 1 TPS on average) and, just like reality, over a sufficient period of time they will eventually tend towards `now`.
